File: java/house/src/main/java/House.java
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class House {

    private static final String[] SUBJECTS = {
        "the house that Jack built.",
        "the malt",
        "the rat",
        "the cat",
        "the dog",
        "the cow with the crumpled horn",
        "the maiden all forlorn",
        "the man all tattered and torn",
        "the priest all shaven and shorn",
        "the rooster that crowed in the morn",
        "the farmer sowing his corn",
        "the horse and the hound and the horn"
    };

    private static final String[] VERBS = {
        "lay in",
        "ate",
        "killed",
        "worried",
        "tossed",
        "milked",
        "kissed",
        "married",
        "woke",
        "kept",
        "belonged to"
    };

    String verse(int verse) {
        if (verse < 1 || verse > SUBJECTS.length) {
            throw new IllegalArgumentException("Verse number must be between 1 and " + SUBJECTS.length);
        }

        StringBuilder sb = new StringBuilder();
        sb.append("This is ");
        sb.append(SUBJECTS[verse - 1]);

        for (int i = verse - 2; i >= 0; i--) {
            sb.append(" that ");
            sb.append(VERBS[i]);
            sb.append(" ");
            sb.append(SUBJECTS[i]);
        }

        return sb.toString();
    }

    String verses(int startVerse, int endVerse) {
        if (startVerse < 1 || endVerse > SUBJECTS.length || startVerse > endVerse) {
             throw new IllegalArgumentException("Invalid verse range");
        }
        return IntStream.rangeClosed(startVerse, endVerse)
                .mapToObj(this::verse)
                .collect(Collectors.joining("\n"));
    }

    String sing() {
        return verses(1, SUBJECTS.length);
    }
}

File: javascript/house/house.js
//
// This is only a SKELETON file for the 'House' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class House {
  static #subjects = [
    'the house that Jack built.',
    'the malt',
    'the rat',
    'the cat',
    'the dog',
    'the cow with the crumpled horn',
    'the maiden all forlorn',
    'the man all tattered and torn',
    'the priest all shaven and shorn',
    'the rooster that crowed in the morn',
    'the farmer sowing his corn',
    'the horse and the hound and the horn',
  ];

  static #actions = [
    null,
    'that lay in',
    'that ate',
    'that killed',
    'that worried',
    'that tossed',
    'that milked',
    'that kissed',
    'that married',
    'that woke',
    'that kept',
    'that belonged to',
  ];

  static verse(n) {
    const lines = [];
    const idx = n - 1;
    lines.push('This is ' + House.#subjects[idx]);
    for (let i = idx; i > 0; i--) {
      lines.push(`${House.#actions[i]} ${House.#subjects[i - 1]}`);
    }
    return lines;
  }

  static verses(start, end) {
    const allVerses = [];
    for (let v = start; v <= end; v++) {
      allVerses.push(...House.verse(v));
      if (v !== end) {
        allVerses.push('');
      }
    }
    return allVerses;
  }
}
